### ISSUE TITLE
Allow monitor plugins to emit raw logs

### ISSUE THREAD

**Issue by imron** [scalyr/scalyr-agent-2#21]
From email:
> It would be nice, though not immediately essential, if we came out of this with a mechanism for emitting raw logs that was simple enough that we could document it for external use.

Perhaps allowing a log format string in the config file that conformed to what Python uses for [log records](https://docs.python.org/2/library/logging.html#logrecord-attributes)?

**Comment by steve-scalyr** [comment]
What purpose would the format string serve?

I was thinking of this as a tool for users who want to import messages from
some other logging system -- as you're doing with syslog. Typically, these
messages will already be fully formatted, and we'll just want to copy them
byte-for-byte. Unless I'm missing something?

**Comment by czerwingithub** [comment]
I can see it both ways.  We may need a format string when folks wish to write to a log file directly in the agent using the normal Python logger class.  When they call logger.info("Hi") vs logger.error("Hi"), it is only the Python format string that adds in the appropriate text to represent the error level.

For this particular feature (syslog), we are getting lines that already contain all of the information, so there's no need for a format string.

In deference to the fact we wish python developers to feel at home, I do think we should eventually expose the format string, but we do not necessarily have to do it in this pass.

There are multiple ways we can skin this cat.  Imron, let me know which of these you favor:

1.  Extend the log_config instance variable on a ScalyrMonitor to allow you to specify multiple log files that should be copied to the server for this monitor.  The easiest way to do that is for log_config to either be a list of JsonObjects or a single JsonObject (where the JsonObject contains a single path's configuration)

This would allow the monitor developers to just create totally independent Logger and Handler instances and do with them what they want.  If they want to get an error into the agent.log, they can still use the monitor instance's _logger variable.  This is probably the most flexible option and the closest to traditional Python approach.

The drawback is that there will be two files generated by the monitor.  Of course, one file will be empty if they do not emit any metrics or use it in any way -- and that makes it even more confusing to the developer since they still have to create a config for something they won't use.

Also, we would probably want to expose some of our Logger filters for them to use, such as the rate filter which limits how many bytes can be written to the log per second -- which helps make sure user's Scalyr log volume quota doesn't get gobbled up by a run away/defective monitor implementation.

2.  Allow a ScalyrMonitor to indicate that its logger instance and handler should not have the Scalyr extensions -- no routing errors to agent.log, no metric logging, etc.  The logger instance is just a regular Logger instance (though maybe with our rate limiter filters to make sure the monitor doesn't send too many bytes to the log file).  We do not even add in a handler for it.  The monitor can then set whatever format string they want, etc.  Of course, they need to remember to add the handler and set the format string they want.

This might be better because the common case is the monitor developer really only wants one log -- there is just this mode where they can say "No, I really just want it to be a naked Logger instance without all of your stuff."

3.  Add in a new optional argument for all log methods on a metric logger called "raw" where if it is True, the message is always emitted directly the the metric log with no format.

I already went down the route of having a "emit_to_metric_log" argument on the Logger instances similar to this, so this is just an extension of what is already there.  

This mainly has the advantage that it is the easiest to implement and less disruptive for the monitor developers to use.

Thoughts?

**Comment by imron** [comment]
I think that option 1 is probably more complexity than we need, unless your users are going to want multiple logs for a given plugin (I don't know enough about this to comment).

I think 3 is more or less what I meant when I said provide access to the format string, except instead of passing a 'raw' flag, we'd be passing the format string.  At the moment, the format string for metric logs is always fixed at ```%(asctime)s [%(monitor_name)s] %(message)s```.  By exposing the format string in the config file, someone could just set it ```%(message)s``` and then they'd just get the raw log message.  It's basically a 'raw' flag, but with the flexibility that someone could add other information from the log record if they wanted or needed to.

The other option I could see is sort of a mix of 2 and 3, which would be to provide some way for developers to specify a handler and formatter for the metric log - perhaps a factory method in the ScalyrMonitor super class ( _create_log_handler and _create_log_formatter) lthat returns our regular handler and formatter by default, and if the developer of a plugin wants to provide their own, they can just override the method to return the one they want.

I guess it depends on what Steve meant by 'external use' in his original comment and whether it's going to be people who just want to set a config option for existing monitors, or if it's going to be developers making their own plugins.

**Comment by steve-scalyr** [comment]
I meant people making their own plugins. I wouldn't see a need to customize
the log format of a plugin you didn't write yourself.

**Comment by imron** [comment]
That being the case, I think allowing the user to provide a handler and/or formatter would be a good choice, either through overriding a method or through some other mechanism.

We could just document the way I'm currently doing it for the syslog monitor (overriding open_metric_log and creating my own handler there), however that doesn't (currently) get the rate logging filters applied (that's probably something I should add), and it's maybe not a good idea to just document that users should do that, and then trust that they'll follow that instruction.

**Comment by czerwingithub** [comment]
Yes, I would suggest we go with the option of ScalyrMonitor having a _create_log_handler class that can be overridden by the derived monitors.  I don't think we need a _create_log_formatter method since the log formatter is a property of the log handler... you can set whatever formatter you want in _create_log_handler.

If you are up for it, you could do this bit of refactoring yourself and add it to the pull request.

I think we should make a point of using the rate logging filters from the get go.  It might make sense to make those classes more available for monitor developers to use.  The only problem is that I, at the time, could not come up with a clean way of preventing serializing the log record twice -- so I had to split the functionality between BaseFormatter and RateLimitLogFilter with a dependency between them.

Imron, if you are up for it, why don't you take a pass at BaseFormatter and RateLimitLogFilter and see if there's a better way to implement them.   I would like to avoid serializing the log record twice if we can, but maybe it doesn't cost that much anyway.  Anyway, if we are going to have other people using the abstraction, it would be good to get a second pair of eyes on it to see if there's  a better interface.

**Comment by imron** [comment]
>  I don't think we need a _create_log_formatter method since the log formatter is a property of the log handler... you can set whatever formatter you want in _create_log_handler.

My thinking was for the case when the user doesn��t want a different handler, but they do want a different formatter.  With a _create_log_formatter, they don��t have to worry about setting up their own handler just to get a different format.

> If you are up for it, you could do this bit of refactoring yourself and add it to the pull request.

Will do.

> Imron, if you are up for it, why don't you take a pass at BaseFormatter and RateLimitLogFilter and see if there's a better way to implement them. I would like to avoid serializing the log record twice if we can, but maybe it doesn't cost that much anyway. 

No problem.

**Comment by czerwingithub** [comment]
On Wed, Apr 15, 2015 at 10:45 PM, Imron <notifications@github.com> wrote:

>
> > I don't think we need a _create_log_formatter method since the log
> formatter is a property of the log handler... you can set whatever
> formatter you want in _create_log_handler.
>
> My thinking was for the case when the user doesn��t want a different
> handler, but they do want a different formatter. With a
> _create_log_formatter, they don��t have to worry about setting up their own
> handler just to get a different format.
>

Ok, that's fine with me.
